packages/typespec-client-generator-core/test/decorators/legacy-hierarchy-building.test.ts

```diff
@@ -0,0 +1,535 @@
+  strictEqual(kingSalmonPropertiesModel.properties.length, 2);
+  strictEqual(kingSalmonPropertiesModel.properties[0].name, "farmed");
+  strictEqual(kingSalmonPropertiesModel.properties[0].type.kind, "constant");
+  strictEqual(kingSalmonPropertiesModel.properties[0].type.value, false);
+  strictEqual(kingSalmonPropertiesModel.properties[1].name, "size");
+  strictEqual(kingSalmonPropertiesModel.properties[1].type.kind, "int32");
+  strictEqual(salmonModel.properties.length, 2);
+  strictEqual(salmonModel.properties[0].name, "kind");
+  strictEqual(salmonModel.properties[0].type.kind, "string");
+  strictEqual(salmonModel.properties[1].type, salmonPropertiesModel);
+  // Verify .discriminatedSubtypes is correctly populated
+  ok(salmonModel.discriminatedSubtypes);
+  strictEqual(salmonModel.discriminatedSubtypes["kingsalmon"], kingSalmonModel);
+  strictEqual(salmonModel.discriminatedSubtypes["smartkingsalmon"], smartKingSalmonModel);
+  ok(kingSalmonModel.discriminatedSubtypes);
+  strictEqual(kingSalmonModel.discriminatedSubtypes["smartkingsalmon"], smartKingSalmonModel);
+  ok(!smartKingSalmonModel.discriminatedSubtypes);
+});
+
+it("circular inheritance", async () => {
+  await runner.compile(`
+      @service
+      namespace TestService;
+
+      model A extends B {
+        propA: string;
+      }
+
+      @Legacy.legacyHierarchyBuilding(A)
+      model B extends C {
+        propB: string;
+      }
+
+      @usage(Usage.input)
+      model C {
+        propC: string;
+      }
+
+      @route("/test")
+      op test(): A;
+    `);
+
+  const modelA = runner.context.sdkPackage.models.find((m) => m.name === "A");
+  const modelB = runner.context.sdkPackage.models.find((m) => m.name === "B");
+  const modelC = runner.context.sdkPackage.models.find((m) => m.name === "C");
+
+  ok(modelA);
+  ok(modelB);
+  ok(modelC);
+
+  strictEqual(modelA.baseModel?.name, "B");
+  strictEqual(modelB.baseModel?.name, "A");
+  strictEqual(modelC.baseModel, undefined);
+
+  strictEqual(modelA.properties.length, 1);
+  strictEqual(modelA.properties[0].name, "propA");
+  strictEqual(modelB.properties.length, 1);
+  strictEqual(modelB.properties[0].name, "propB");
+  strictEqual(modelC.properties.length, 1);
+  strictEqual(modelC.properties[0].name, "propC");
+});
+
+it("conflicting inheritance", async () => {
+  const [_, diagnostics] = await runner.compileAndDiagnose(`
+      @service
+      namespace TestService;
+
+      model A {
+        propA: string;
+      }
+
+      model B {
+        propB: string;
+      }
+
+      @Legacy.legacyHierarchyBuilding(B)
+      model C extends A {  // Doesn't have propB but @Legacy.legacyHierarchyBuilding says B
+        propC: string;
+      }
+
+      @route("/test")
+      op test(): C;
+    `);
+
+  // Should warn about conflicting inheritance
+  expectDiagnostics(diagnostics, {
+    code: "@azure-tools/typespec-client-generator-core/legacy-hierarchy-building-conflict",
+  });
+});
+
```

Comment:
One case I'd like to confirm is if A extends B extends C, which means A could only override properties with narrow scopes e.g only allow to override one optional property to required. If with legacyHierarchyBuilding decorator do we gurantee that?

```
@discriminator("kind")
model C {
    kind: string;
}

model A extends B {
      propA: string;
}
@Legacy.legacyHierarchyBuilding(A)
model B extends C {
  propB: string;
  // override A's property
  propA?: string;
}
```

Reply:
Okay, not sure we have any case which would requires more narrow scopes in swagger side? 

The `fail` here means a tcgc error?